cub/agent/agent_segmented_radix_sort.cuh

```diff
@@ -0,0 +1,294 @@
+/******************************************************************************
+ * Copyright (c) 2011-2021, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include "../config.cuh"
+#include "../util_type.cuh"
+#include "../util_namespace.cuh"
+#include "../block/block_radix_sort.cuh"
+#include "../agent/agent_radix_sort_upsweep.cuh"
+#include "../agent/agent_radix_sort_downsweep.cuh"
+
+CUB_NAMESPACE_BEGIN
+
+
+template <
+  bool      IS_DESCENDING,                  ///< Whether or not the sorted-order is high-to-low
+  typename  SegmentedPolicyT,               ///< Chained tuning policy
+  typename  KeyT,                           ///< Key type
+  typename  ValueT,                         ///< Value type
+  typename  OffsetT>                        ///< Signed integer type for global offsets
+struct AgentSegmentedRadixSort
```

Comment:
It's implementing `DeviceSegmentedRadixSort`.  `DeviceSegmentedSort` relies on this class for large/medium segments only. To call it `AgentSegmentedSort ` I'd also need to integrate `AgentSubWarpSort` into it. `AgentSubWarpSort` is used in separate kernels, I wouldn't like to mix those two.

Reply:
> It's implementing `DeviceSegmentedRadixSort`

`DeviceSegmentRadixSort` is not using this agent:

https://github.com/senior-zero/cub/blob/main-feature/github/seg_sort/cub/device/dispatch/dispatch_radix_sort.cuh#L397-L404

Reply:
That sounds reasonable, thanks for the context!

To avoid confusion in the meantime, could you file a bug proposing that we switch to in-kernel loops with the details from your initial investigation, and leave a comment on the class's implementation that refers to this bug?